prow/gcs/reporter/reporter.go

```diff
@@ -0,0 +1,213 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reporter
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"path"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/GoogleCloudPlatform/testgrid/metadata"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/api/googleapi"
+	"k8s.io/test-infra/prow/errorutil"
+
+	prowv1 "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/gcsupload"
+	"k8s.io/test-infra/prow/pod-utils/downwardapi"
+)
+
+const reporterName = "gcsreporter"
+
+type gcsReporter struct {
+	cfg     config.Getter
+	dryRun  bool
+	logger  *logrus.Entry
+	storage *storage.Client
+}
+
+func (gr *gcsReporter) Report(pj *prowv1.ProwJob) ([]*prowv1.ProwJob, error) {
+	ctx, _ := context.WithTimeout(context.Background(), 10*time.Second) // TODO: pass through a global context?
+
+	_, _, err := gr.getJobDestination(pj)
+	if err != nil {
+		gr.logger.Infof("Not uploading %q (%s#%s) because we couldn't find a destination: %v", pj.Name, pj.Spec.Job, pj.Status.BuildID, err)
+		return []*prowv1.ProwJob{pj}, nil
+	}
+	stateErr := gr.reportJobState(ctx, pj)
+	prowjobErr := gr.reportProwjob(ctx, pj)
+
+	return []*prowv1.ProwJob{pj}, errorutil.NewAggregate(stateErr, prowjobErr)
+}
+
+func (gr *gcsReporter) reportJobState(ctx context.Context, pj *prowv1.ProwJob) error {
+	if pj.Status.State == prowv1.TriggeredState {
+		return gr.reportStartedJob(ctx, pj)
+	}
+	if pj.Complete() {
+		return gr.reportFinishedJob(ctx, pj)
+	}
+	return nil
+}
+
+// reportStartedJob uploads a started.json for the job, iff one did not already exist.
```

Comment:
This is not true. Triggered is the initial state -- https://github.com/kubernetes/test-infra/blob/b82ca85d511832330bb525a56c354dbef65bb804/prow/pjutil/pjutil.go#L56

Plank and crier will see this state at roughly the same time. At this point plank hasn't even created the pod, which means there's zero chance its assigned to a node, downloaded images, and completed the clone and initupload init containers.

Is there a reason we only wait for the pod to complete for finished.json, why not also started.json?

Reply:
This _is_ true - but we can expect under any reasonable circumstances that circumstances it will upload one (and then podutils will overwrite it). The precondition is purely to defend against unreasonable circumstances (bugs or very weird infra issues).

I chose to create started.json before the job starts to ensure GCS reflects an accurate status, both in the spirit of crier reporting events as they happen and so that tools that read GCS can consistently see that the job exists immediately.

That said, I'm not strongly attached to this - is there a reason it would be better to wait?

Reply:
So long as all the other utilities will overwrite this file then this seems fine, maybe just change the comment? This will write the first started.json file basically 100% of the time.

Reply:
Done.